include/zephyr/sys/async.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Public APIs for Async utilities
+ *
+ * Functions and datatypes useful for safely communicating to a thread or work
+ * queue of an asynchronous functions completion.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ASYNC_H_
+#define ZEPHYR_INCLUDE_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+/**
+ * @brief Async operation callback type
+ *
+ * Common function pointer type used for asynchronous completion callbacks
+ * and implemented for some commonly used IPC notification mechanisms.
+ *
+ * May also be implemented by an application with the understanding
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
+
+#ifdef CONFIG_ASYNC
+
+/**
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
+	/** Semaphore that will be given on completion */
+	struct k_sem *sem;
+
+	/** Result of async call */
+	int result;
+};
+
+/**
+ * Asynchronous semaphore completion callback
+ *
+ * Sets the result and gives the semaphore
+ *
+ * @param result Result to assign to the async_sem
+ * @param userdata Pointer to a struct async_sem
+ */
+void async_semaphore_cb(int result, void *userdata);
+
+/**
+ * Signal completion callback
+ *
+ * Sets the result and raises a k_poll_signal
+ *
+ * @param result Result to assign to the signal
+ * @param userdata Pointer to a struct k_poll_signal
+ */
+void async_signal_cb(int result, void *userdata);
+
+/**
+ * Mutex completion object with return code.
+ */
+struct async_mutex {
+	/** Mutex that will be unlocked on  completion */
+	struct k_mutex *mutex;
+
+	/** Result of async call */
+	int result;
+};
+
+/**
+ * Asynchronous mutex completion callback
+ *
+ * Sets the result and unlocks the mutex
+ *
+ * @param result Result to assign to async_mutex
+ * @param userdata Pointer to struct async_mutex
+ */
+void async_mutex_cb(int result, void *userdata);
+
+/**
+ * Work completion object with return code.
+ */
+struct async_work {
+	/** Work that will be submitted on completion */
+	struct k_work *work;
+
+	/** Result of async call */
+	int result;
+};
+
+/**
+ * Asynchronous work completion callback
+ *
+ * Submits an associated k_work task to the global work queue
+ *
+ * @param result Result to assign to async_work
+ * @param userdata Pointer to struct async_work
+ */
+void async_work_cb(int result, void *userdata);
```

Comment:
Another useful helper would be for submitting work to a specific queue. The queue could be specified as a field in `struct async_work`. It could even be generalized as to always take a queue in `struct async_work` and use the global work queue if `NULL`.